Add MACD as an indicator

daveln opened the pull request:
Add MACD as an indicator

Also requested in
https://forum.portfolio-performance.info/t/chartanalyse-charttools-und-indikatoren/744/7 and https://forum.portfolio-performance.info/t/chart-analyses-in-stock-graph-like-rsi-or-macd-or-others/21484

daveln commented:
@buchen 
Potential issues:
* I could modify TimelineChart.getOrCreateAxis to accept only SecuritiesChart.ChartDetails but I'm not decided if a defined type is required for the key. The key could be a 'new Object()' and it would still work. For now I think it is ok to not define the key type.
* The color is just a few ints that were already used as color consts in a unique combination. I don't know if you have a preference to change that to something else.
* No translations other than English and Dutch.
* The test is modified from SimpleMovingAverageTest. I could fold testSufficientPriceDataPass and testSufficientPriceDataStartDate into one test. Eh.

Thoughts?

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,112 @@
+        double subtrahendEma = minuendEma;
+        double signalEma = 0;
+
+        for (int i = 0; i < prices.size(); i++) // NOSONAR
```
Change to... 
for (SecurityPrice price : prices) {
    LocalDate date = price.getDate();
    // ...
}

daveln replied:
Fair. Changed, squashed, force pushed.

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,121 @@
+            return;
+
+        List<SecurityPrice> prices = security.getPricesIncludingLatest();
+        if (prices == null)
+            return;
```
```suggestion
        if (prices == null || prices.isEmpty())
            return;
```

daveln replied:
Fair. I've determined that security.getPricesIncludingLatest() can never return null. I'm dropping the null check, and replacing with your suggested prices.isEmpty().

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,121 @@
+
+    public MovingAverageConvergenceDivergence(Security security, ChartInterval interval)
+    {
+        this.security = security;
```
```suggestion
        this.security = Objects.requireNonNull(security);
```

daveln replied:
Changing as suggested.

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,121 @@
+
+    private void calculateMacd()
+    {
+        if (security == null)
+            return;
```
```suggestion
```

daveln replied:
Changing as suggested.

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,121 @@
+        int index = Collections.binarySearch(prices, new SecurityPrice(interval.getStart(), 0),
+                        new SecurityPrice.ByDate());
+
+        if (index < 0)
+            index = -index - 1;
+
+        if (index >= prices.size())
+            return;
```
```suggestion
        index = Math.max(0, index);
```

daveln replied:
Hmm. In the previous code this index was only used a stop condition (if the interval was in the future of the prices list). I've refactored this and made it more explicit what the intent is.

Nirus2000 commented on name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java:
```diff
@@ -0,0 +1,121 @@
+            datesMacd.add(date);
+        }
+
+        Date[] timelineDates = TimelineChart.toJavaUtilDate(datesMacd.toArray(new LocalDate[0]));
+        macd.setDates(timelineDates);
```
```suggestion
        macd.setDates(TimelineChart.toJavaUtilDate(datesMacd.toArray(new LocalDate[0])));
        signalLine.setDates(macd.getDates());
```

daveln replied:
After changing the calculation for correctness, this code was changed; MACD and MACDSignal have different start dates. In future there will be a xlsx file with the full MACD and MACDSignal calculation.

daveln commented:
* Added formatting to the MACD/MACDSignal values in the SecuritiesChart/TimelineChartToolTip.
* Calculation revisited after noticing that the start of the series was deviating from correct values. (The end of the series was correct.)
* Test is completely reworked. The test now has full coverage of MovingAverageConvergenceDivergence class.
* Added an excel file with matching csv that contains the suitable calculations for MACD, RSI, EMA, SMA for a security.
* It is used in the test to validate calculation correctness for MACD.
* Processed review comments.

buchen commented:
> I could modify TimelineChart.getOrCreateAxis to accept only SecuritiesChart.ChartDetails but I'm not decided if a defined type is required for the key. The key could be a 'new Object()' and it would still work. For now I think it is ok to not define the key type.

I think that is okay. The ChartDetails key should not leak into TimelineChart (which is used in other locations as well).

Anyway, the SecurityChart class has grown over time as more and more things have been bolted to it. At one point, we'll have to refactor this class. But for MACD it is okay to add it the existing way.

I'll add some additional translations and will merge it. Thanks.
